normandy/recipes/signing.py

```diff
@@ -190,9 +193,26 @@ def verify_x5u(url, expire_early=None):
     chain. Otherwise, raise an exception explaining why they are not
     valid.
     """
-    req = requests.get(url)
-    req.raise_for_status()
-    pem = req.content.decode()
+    cache_key = f"fetch-x5u-pem::{url}"
+
+    pem = None
+    if settings.X5U_CACHE_TIME:
+        pem = cache.get(cache_key)
+
+    if isinstance(pem, Exception):
+        raise CachedRequestException(url=url)
+
+    if pem is None:
+        try:
+            resp = requests.get(url, timeout=X5U_REQUEST_TIMEOUT)
+            resp.raise_for_status()
+        except requests.RequestException as exc:
+            cache.set(cache_key, exc, settings.X5U_CACHE_TIME)
```

Comment:
I'm nervous about caching failures for as long as we cache successes. This means a transient error or latency spike could cause all heartbeats requests to this process to fail for 10 minutes. It's common to have separate knobs for successes vs errors (e.g. [cloudfront does](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/custom-error-pages-expiration.html)). I'd lean towards making the error's TTL closer to the duration of the heartbeat request; that's currently averaging ~4 seconds.